samza-core/src/main/java/org/apache/samza/container/host/PosixCommandBasedStatisticsGetter.java

```diff
@@ -44,20 +47,71 @@ public class PosixCommandBasedStatisticsGetter implements SystemStatisticsGetter
   private List<String> getAllCommandOutput(String[] cmdArray) throws IOException {
     log.debug("Executing commands {}", Arrays.toString(cmdArray));
     Process executable = Runtime.getRuntime().exec(cmdArray);
-    BufferedReader processReader;
     List<String> psOutput = new ArrayList<>();
 
-    processReader = new BufferedReader(new InputStreamReader(executable.getInputStream()));
-    String line;
-    while ((line = processReader.readLine()) != null) {
-      if (!line.isEmpty()) {
-        psOutput.add(line);
+    try (BufferedReader processReader = new BufferedReader(new InputStreamReader(executable.getInputStream()));
+         BufferedReader errorReader = new BufferedReader(new InputStreamReader(executable.getErrorStream()))) {
+
+      // Read output stream
+      String line;
+      while ((line = processReader.readLine()) != null) {
+        if (!line.isEmpty()) {
+          psOutput.add(line);
+        }
+      }
+
+      // Consume error stream to prevent blocking
+      consumeErrorStream(errorReader, cmdArray);
+
+      // Wait for the process to complete to prevent resource leak
+      try {
+        boolean finished = executable.waitFor(COMMAND_TIMEOUT_SECONDS, TimeUnit.SECONDS);
```

Comment:
What will be the behavior if the executable did not complete within the timeout?